test/hotspot/jtreg/testlibrary_tests/template_framework/tests/TestTemplate.java

```diff
@@ -0,0 +1,1269 @@
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8344942
+ * @summary Test some basic Template instantiations. We do not necessarily generate correct
+ *          java code, we just test that the code generation deterministically creates the
+ *          expected String.
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib /
+ * @run main template_framework.tests.TestTemplate
+ */
+
+package template_framework.tests;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.HashSet;
+
+import compiler.lib.template_framework.Template;
+import compiler.lib.template_framework.FilledTemplate;
+import compiler.lib.template_framework.UnfilledTemplate;
+import compiler.lib.template_framework.Name;
+import compiler.lib.template_framework.Hook;
+import compiler.lib.template_framework.TemplateBinding;
+import compiler.lib.template_framework.RendererException;
+import static compiler.lib.template_framework.Template.body;
+import static compiler.lib.template_framework.Template.$;
+import static compiler.lib.template_framework.Template.let;
+import static compiler.lib.template_framework.Template.fuel;
+import static compiler.lib.template_framework.Template.setFuelCost;
+import static compiler.lib.template_framework.Template.addName;
+import static compiler.lib.template_framework.Template.weighNames;
+import static compiler.lib.template_framework.Template.sampleName;
+
+public class TestTemplate {
+    interface FailingTest {
+        void run();
+    }
+
+    private record MyPrimitive(String name) implements Name.Type {
+        @Override
+        public boolean isSubtypeOf(Name.Type other) {
+            return other instanceof MyPrimitive(String n) && n == name();
+        }
+
+        @Override
+        public String toString() { return name(); }
+    }
+    private static final MyPrimitive myInt = new MyPrimitive("int");
+    private static final MyPrimitive myLong = new MyPrimitive("long");
+
+    public static void main(String[] args) {
+        testSingleLine();
+        testMultiLine();
+        testBodyTokens();
+        testWithOneArguments();
```

Comment:
Another small glitch:
```suggestion
        testWithOneArgument();
```